### ISSUE TITLE
skipawait per-operation (create, update, delete)

### ISSUE THREAD

**Issue by antdking** [pulumi/pulumi-kubernetes#2551]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

We have a case where deletions of a resource are slow; around 1 hour (GCP/GKE issue, not Pulumi).
Using `skipAwait` means we won't wait fully for creation.

It would be good to be able to have `skipAwait` for deletion only. In our case, it doesn't matter if the resource is left around for a while, as long as it's been marked for termination.

**Comment by antdking** [comment]
details of the problem is that GKE Service Network Endpoint Groups don't cleanup timely, which slows down Namespace deletion beyond a reasonable timeout.

For now, we're applying the `skipAwait` annotation using a `NamespacePatch`, so that we still await on creation, but don't await for deletion. As it's a Namespace, we rarely update the resource, so there's no real downside outside of code overhead.

**Comment by blampe** [comment]
@antdking I'm curious how you would expect the UX to work around this? I can imagine a couple options:

1. Introduce some additional annotations like `skipAwaitCreate`, `skipAwaitDelete`, `skipAwaitUpdate`.
2. Modify the existing annotation to allow specifying specific operations like `"skipAwait": "create,delete"`, in addition to still supporting all operations with `"skipAwait": "true"`.

My preference is probably (2) but I'm curious what you think. Or did you have something else in mind?

**Comment by antdking** [comment]
(2) seems fine given the limitations in k8s metadata.

The main concern I see with it is transformer usecases.

In most cases, they'll just want to append, so it'd be worth allowing strings like "true,delete" or "delete,delete"; where the resource decides to always skip, but the transformer is only concerned about "create" being skipped.

Would it be better to expose as a list of strings in the sdk, or keep it as the comma separated list to keep the sdk aligned with the underlying data?

**Comment by blampe** [comment]
Question for @antdking, @oktocat and others: do we need a distinction between creates and updates, or are we really just interested in something like `skipAwaitDelete` to compliment `skipAwait`?

In other words, is there a use case where you would want `skipAwaitCreate` but _not_ `skipAwaitUpdate`? I can't really think of an instance where that would be useful because we really only care about "readiness," and that doesn't care how we manipulated the manifest.

Our current logic essentially waits on create/updates the same, which feels "right" and analogous to invoking `kubectl wait --for` after an `apply` -- `wait` doesn't care how the thing got there, only that it becomes ready

**Comment by oktocat** [comment]
@blampe I agree with you, it feels more intuitive to have the same-ish behaviour as `kubectl wait`. Especially if that's explicitly called out in the option's docstring.

**Comment by antdking** [comment]
having delete or ready makes sense, I can't think of a specific example where creation and updating needs to be treated differently

**Comment by blampe** [comment]
@antdking @oktocat FYI we do have an (undocumented) `pulumi.com/deletionPropagationPolicy` annotation which can take values of `orphan` or `background`. Pulumi defaults to [foreground](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#foreground-deletion) deletion, which waits for all dependents to be cleaned up.

@antdking for your namespace example, I do wonder if using a "background" policy would do the trick here, allowing the ~1hr deletion to happen out of band. Could you give that a try?

**Comment by blampe** [comment]
We've discussed this a bit and decided it's probably a bug that `skipAwait` affects deletion at all. Given that the `pulumi.com/deletionPropagationPolicy` annotation exists as a workaround for slow deletes, it's unlikely we'll be expanding or changing the behavior of `skipAwait` until we've thought through #1232 and #3154.
